<!doctype html>
<html>
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <title>RNASkim by zzj</title>

    <link rel="stylesheet" href="stylesheets/styles.css">
    <link rel="stylesheet" href="stylesheets/pygment_trac.css">
    <meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">
    <!--[if lt IE 9]>
    <script src="//html5shiv.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->
  </head>
  <body>
    <div class="wrapper">
      <header>
        <h1>RNASkim</h1>
        <p>RNA-Skim: a rapid method for RNA-Seq quantification at transcript level</p>

        <p class="view"><a href="https://github.com/zzj/RNASkim">View the Project on GitHub <small>zzj/RNASkim</small></a></p>


        <ul>
          <li><a href="https://github.com/zzj/RNASkim/zipball/master">Download <strong>ZIP File</strong></a></li>
          <li><a href="https://github.com/zzj/RNASkim/tarball/master">Download <strong>TAR Ball</strong></a></li>
          <li><a href="https://github.com/zzj/RNASkim">View On <strong>GitHub</strong></a></li>
        </ul>
      </header>
      <section>
        <h1>
<a name="rnaskim" class="anchor" href="#rnaskim"><span class="octicon octicon-link"></span></a>RNASkim</h1>

<p>RNA-Skim: a rapid method for RNA-Seq quantification at transcript level</p>

<h2>
<a name="clustering-the-transcripts-based-on-their-sequence-similarities" class="anchor" href="#clustering-the-transcripts-based-on-their-sequence-similarities"><span class="octicon octicon-link"></span></a>Clustering the transcripts based on their sequence similarities.</h2>

<p>TODO(zzj)</p>

<h2>
<a name="building-the-index-for-a-given-transcript-clustering-result" class="anchor" href="#building-the-index-for-a-given-transcript-clustering-result"><span class="octicon octicon-link"></span></a>Building the index for a given transcript clustering result</h2>

<p>This command downloads a clustering result of the mouse transcriptome. </p>

<pre>
wget http://www.csbio.unc.edu/rs/clustered_gene.fa
</pre>

<p>The clustered_gene.fa file uses fasta format, but it may contain more than one transcript per two lines, for example:</p>

<pre>&gt;gid|tid1|tid2
ATCTATTA|TACTAATC
</pre> 

<p>This presents a transcript cluster named gid which contains two transcripts tid1 and tid2. And the sequence of tid1 is ATCTATTA and the sequence of tid2 is TACTAATC. "|" is the delimiter for both the sequence ids and the sequences. And the first id of the sequence ids is the gene id.</p>

<p>Now you can run the following command to find all sig-mer regions.</p>

<pre>./rs_index -transcript_fasta=clustered_gene.fa -index_file=clustered_gene.fa.pb -num_threads 4</pre>

<p>A sig-mer region is a sequence that all k-mers from the region are sig-mers of the transcript cluster. (Check out the document for the GeneSignatures class at rnasigs.proto for more details). The clustered_gene.fa.pb contains the corresponding GeneSignatures class for every transcript cluster. The storage format is at <a href="https://github.com/zzj/rnasigs/wiki/Code-Structure#wiki-storage-proto_datah">Storage format for protobuf in RNA-Skim</a>.</p>

<p>The default sig-mer size is 40. </p>

<pre>./rs_index -transcript_fasta=clustered_gene.fa -index_file=clustered_gene.fa.pb -rs_length 50 -num_threads 4</pre>

<p>You can check all arguments information by</p>

<pre>./rs_index --help </pre>

<p><strong>The "--help" parameter also works for other executables.</strong> The output list all parameters into different sections by their source files, and only the parameters defined in the RNA-Skim sources (the file names start with "rs_") are defined by RNA-Skim. Or you can run </p>

<pre>./rs_index --helpshort
./rs_select --helpshort
./rs_count --helpshort
./rs_estimate --helpshort
</pre>

<p>to filter out all unnecessary parameters.</p>

<h2>
<a name="selecting-sig-mers-from-the-sig-mer-regions" class="anchor" href="#selecting-sig-mers-from-the-sig-mer-regions"><span class="octicon octicon-link"></span></a>Selecting sig-mers from the sig-mer regions</h2>

<p>OK, now you have the clustered_gene.fa.pb file. Let's select sig-mers from all sig-mer regions:</p>

<pre>./rs_select -index_file=clustered_gene.fa.pb -selected_keys_file=clustered_gene.fa.sk</pre>

<p><strong>If you use sig-mer size other than the default value, you should specify the length in the parameter list as well,</strong> for example:</p>

<pre>./rs_select -index_file=clustered_gene.fa.pb -selected_keys_file=clustered_gene.fa.sk  -rs_length 50</pre>

<p>Please be careful about this. This is annoying and should be <a href="https://github.com/zzj/rnasigs/issues/14">fixed</a>. </p>

<p>You may see some warnings like this,</p>

<pre>E0309 21:20:25.820291 1990034192 rs_select.cc:130] ENSMUST00000114890 has 1 rna_signatures. Skipped.</pre>

<p>For some transcripts, RNA-Skim cannot get enough number of sig-mers, so it does not quantify such transcripts because the results of those transcripts are not reliable. Most of such transcripts are either too short (less than 150 bps) or categorized as "predicted gene".</p>

<p>The output file (clustered_gene.fa.sk) is a list of SelectedKey objects. For details, please checkout the comments for SelectedKey at rnasigs.proto.</p>

<h2>
<a name="simulating-the-rna-seq-data" class="anchor" href="#simulating-the-rna-seq-data"><span class="octicon octicon-link"></span></a>Simulating the RNA-Seq data</h2>

<p>Please run the download.py file first (<a href="https://github.com/zzj/rnasigs/wiki/Internal-Development-Guide">Internal Development Guide</a>). Remember to clean up the gtf file (details at the internal develop guide)</p>

<p>We use <a href="http://sammeth.net/confluence/display/SIM/Home">Flux simulator</a> to simulate RNA-Seq data. If you have run bootstrap.sh command, it should be installed at lib/flux-simulator-1.2. If not, please file an issue and fix bootstrap.sh file. </p>

<p>The following command generates a test data of 1million paired reads</p>

<pre>cd ..
mkdir test
cp simulation/test.par test
cd test
export FLUX_MEM="6G";
../lib/flux-simulator-1.2/bin/flux-simulator -p test.par
</pre>

<p>Now, we have four extra files (test.bed, test.fastq, test.lib, test.pro). The two most important files are test.fastq (RNASeq fastq files) and the test.pro (the simulated abundance values. More details at <a href="http://sammeth.net/confluence/display/SIM/.PRO+Transcriptome+Profile">here</a>).</p>

<p>This command split the fastq files into two fasta files (test.fastq_1  and test.fastq_2) and two fastq files (test.fastq_1.fastq and test.fastq_2.fastq).</p>

<pre>
python ../simulation/split_fasta.py test.fastq
</pre>

<p>For more details, check out the web site of the flux simulator.</p>

<h2>
<a name="counting-the-sig-mers-for-a-given-rna-seq-dataset" class="anchor" href="#counting-the-sig-mers-for-a-given-rna-seq-dataset"><span class="octicon octicon-link"></span></a>Counting the sig-mers for a given RNA-Seq dataset</h2>

<p>Now, we run rs_count to count all sig-mers stored in the clustered_gene.fa.sk file.</p>

<pre>GLOG_logtostderr=1  ../src/rs_count  -selected_keys_file=clustered_gene.fa.sk -count_file=clustered_gene.fa.cf -read_files1=../test/test.fastq_1 -read_files2=../test/test.fastq_2 -num_threads=1 -rs_length (INPUT_YOUR_SIG_MER_LENGTH_HERE)</pre>

<p>This generates clustered_gene.fa.cf file, which is almost identical with the clustered_gene.fa.sk file, but the count fields in the SelectedKey object in the clustered_gene.fa.cf file is the real occurrences of their corresponding sig-mers.</p>

<h2>
<a name="quantifying-the-abundances-of-transcripts" class="anchor" href="#quantifying-the-abundances-of-transcripts"><span class="octicon octicon-link"></span></a>Quantifying the abundances of transcripts</h2>

<p>This command quantifies the transcriptome based on the counts of sig-mers in the clustered_gene.fa.cf file.</p>

<pre>
../src/rs_estimate -count_file=clustered_gene.fa.cf &gt; estimation
</pre>

<p>There are four columns in the estimation file: transcript id; the length of the transcript; estimated number of relative reads; RPKM value of the transcript.</p>
      </section>
      <footer>
        <p>This project is maintained by <a href="https://github.com/zzj">zzj</a></p>
        <p><small>Hosted on GitHub Pages &mdash; Theme by <a href="https://github.com/orderedlist">orderedlist</a></small></p>
      </footer>
    </div>
    <script src="javascripts/scale.fix.js"></script>
    
  </body>
</html>